typing/ctype.ml

```diff
@@ -263,6 +263,24 @@ let newconstr path tyl = newty (Tconstr (path, tyl, ref Mnil))
 
 let none = newty (Ttuple [])                (* Clearly ill-formed type *)
 
+(**** information for [Typecore.unify_pat_*] ****)
+
+type pattern_environment =
+    { env : Env.t ref;
+      equations_scope : int;
+      allow_recursive_equations : bool; }
+
+let make_pattern_environment env ~equations_scope ~allow_recursive_equations =
+  { env = ref env;
+    equations_scope;
+    allow_recursive_equations; }
+
+let copy_pattern_environment penv =
+  { penv with env = ref !(penv.env) }
+
+let set_equations_scope equations_scope penv =
+  { penv with equations_scope }
```

Comment:
Thank you for the suggestions. I have modified the code to use `module Pattern_env`.

`get_equations_scope` is removed at all, since the safety check inside it has turned out unnecessary after inspecting how the values are set to `equations_scope`.

`set_equations_scope` is kept together with private `Pattern_env.t` to ensure its values
are created only by `Pattern_env.make` or `copy`.